cmake/Modules/Platform/Emscripten.cmake

```
@@ -18,7 +18,13 @@ set(CMAKE_SYSTEM_NAME Emscripten)
 set(CMAKE_SYSTEM_VERSION 1)
 
 set(CMAKE_CROSSCOMPILING TRUE)
-set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS FALSE)
+
+# shared library (side module) support
+set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS TRUE)
+set(CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS "-s SIDE_MODULE=1")    # instead of -shared
+set(CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS "-s SIDE_MODULE=1")  # instead of -shared
+set(CMAKE_STRIP FALSE)  # not supported
+set(BUILD_SHARED_LIBS OFF)  # default to static libs, even if we allow shared ones
```

not needed, default is also static.

---

Our goal should be this table.

| cmake | user option | library type |
|:--:|:--:|:--:|
| `add_library(foo)` | | static |
| `add_library(foo)` | -DBUILD_SHARED_LIBS=ON | dynamic |
| `add_library(foo SHARED)` | | dynamic |
| `add_library(foo)` `option(BUILD_SHARED_LIBS "" ON)` | | **static** |
| `add_library(foo)` `option(BUILD_SHARED_LIBS "" ON)` | -DBUILD_SHARED_LIBS=ON | **dynamic** |

The most important thing will be that `option(BUILD_SHARED_LIBS "" ON)` should be overridden to **OFF** for compatibility unless `BUILD_SHARED_LIBS=ON` is explicitly specified.

Just adding `configure=['cmake', '.', '-DBUILD_SHARED_LIBS=OFF']` will be sufficient for this and `set(BUILD_SHARED_LIBS OFF)` can be deleted.
The other solution will be that caching `BUILD_SHARED_LIBS`, this will overwrite `option(BUILD_SHARED_LIBS "" ON)`.

```suggestion
set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build WebAssembly dynamic libraries. Default value is 'OFF'.") # default to static libs, even if we allow shared ones
```

---

Hi @nokotan,

what you write in this table is very reasonable, but there is a misunderstanding on your side how CMake works.

Yes, `add_library` defaults will be based on the `BUILD_SHARED_LIBS` default. That one is static now.

No, if a user sets either `set(BUILD_SHARED_LIBS ON)` or `option(BUILD_SHARED_LIBS "" ON)` this will be overwritten. This is a clear user decision then (in their `CMakeLists.txt`) and we cannot overwrite this without disabling shared library support altogether (which was done prior to this PR).

I hope this clarifies a bit, please let me know if you have more questions.

---

The question remains, `option(BUILD_SHARED_LIBS "" ON)` seems to do nothing when `BUILD_SHARED_LIBS` is set by user or is already cached, after CMake 3.13. ([CMP0077])

[CMP0077]: https://cmake.org/cmake/help/latest/policy/CMP0077.html#policy:CMP0077

---

That is exactly how all cached variables and options work in CMake.
By default, `BUILD_SHARED_LIBS` is actually not set until we set it here in the toolchain.

Yes, users, toolchain files and cache files can be used by users to overwrite this - that is fine.